### PR TITLE
add user role to JWT token

### PR DESCRIPTION
**Pull request by andrewtookay**

### CODE REVIEW

File: app/dot/cpp/core/services/LoginService.java [hunk 1]
```diff
@@ -127,26 +126,14 @@ private String getAccessToken(String userId) {
             .setSubject(userId)
             .setExpiration(expirationDateAccess)
             .signWith(key)
+            .claim(USER_ROLE, userRole)
```
Comment (sircoderin) [inline app/dot/cpp/core/services/LoginService.java:129]: what does this do?

File: app/dot/cpp/core/services/LoginService.java [hunk 2]
```diff
@@ -156,24 +143,33 @@ private Jws<Claims> getJwsClaims(String jwtToken) throws LoginException {
...
 
-      return user;
-    } catch (BaseException e) {
-      throw LoginException.from(ErrorCodes.USER_NOT_FOUND);
+    if (expirationDate.before(new Date())) {
+      throw LoginException.from(ErrorCodes.EXPIRED_ACCESS);
     }
+
+    final var userRole = claims.get(USER_ROLE, String.class);
+    if (!permittedUserRoles.isEmpty() && !permittedUserRoles.contains(UserRole.valueOf(userRole))) {
```
Comment (sircoderin) [inline app/dot/cpp/core/services/LoginService.java:168]: shouldn't we also add ADMIN to Authentication of editPage and edit from UserController?

Comment (sircoderin) [reply app/dot/cpp/core/services/LoginService.java:168]: I believe setting the permittedUserRoles as default to only USER would be more appropriate, then the check for empty would not be needed. What do you think?

File: app/dot/cpp/core/services/LoginService.java [hunk 3]
```diff
@@ -127,26 +126,14 @@ private String getAccessToken(String userId) {
             .setSubject(userId)
```
Comment (sircoderin) [inline app/dot/cpp/core/services/LoginService.java:126]: How about setting the subject to something as AUTH_TOKENS and inside the claim both userId and userRole. It's just an idea that popped into my head, that way when retrieving them it would be directly by the name

**Review by sircoderin — CHANGES_REQUESTED** [review]
I would have a tendency to refactor some logic from LoginService as it seems a little hard to follow but I can't really be asked to see how to do it so I guess we can leave it for now and maybe refactor if needed

File: app/dot/cpp/core/services/LoginService.java (hunk 1 above)
Comment (andrewtookay) [reply app/dot/cpp/core/services/LoginService.java:129]: Indeed

File: app/dot/cpp/core/services/LoginService.java (hunk 2 above)
Comment (andrewtookay) [reply app/dot/cpp/core/services/LoginService.java:168]: > shouldn't we also add ADMIN to Authentication of editPage and edit from UserController?

They should've had that, that's an issue.

Comment (andrewtookay) [reply app/dot/cpp/core/services/LoginService.java:168]: > I believe setting the permittedUserRoles as default to only USER would be more appropriate, then the check for empty would not be needed. What do you think?

Sure, sounds good.

**Comment by andrewtookay** [comment]
> I would have a tendency to refactor some logic from LoginService as it seems a little hard to follow but I can't really be asked to see how to do it so I guess we can leave it for now and maybe refactor if needed

Sure, some refactor might be useful, but it can be done later.

File: app/dot/cpp/core/services/LoginService.java (hunk 3 above)
Comment (andrewtookay) [reply app/dot/cpp/core/services/LoginService.java:126]: Set the subject "AUTH_TOKENS" for one access token? Why waste a field with a useless text? The way I see it is, this token has this user as the subject, and for this user I have these claims.
